### ISSUE TITLE
improve our sweep function

### ISSUE THREAD

**Issue by gits7r** [spesmilo/electrum#1837]
Currently our sweep function will simply spend everything from an address belonging to a certain private key minus the fee (correctly) calculated by electrum. This is fine for most use cases, but we might improve the function a little bit by making it consider other things as well.

Today an user reported troubles sweeping a private key of an address containing many small (dust) inputs. The transaction got confirmed after > 10 hours, the txid is: ac230aaca7a8a7f9dc2cf4328d479815d3a9332f568293ced0bcdf8e6338bde6

The sweep function should at least check if:
- the inputs swept are not fresh mined coins and have at least n blocks on top of them
- there aren't too many dust inputs in a single transaction (the txid provided above was rejected by node for having too many too small inputs, with some luck eventually got mined)
- if there are many many inputs, don't create a single monster sized transaction, create more instead

**Comment by dabura667** [comment]
https://bitcoinfees.21.co/

Your user paid 9 satoshis / byte as fees, (50000 satoshis / 5503 bytes)

Currently the wait time estimated for that amount is between 15 - 660 minutes.

I am guessing your user tried to manually input the fee by "ballparking" it. (0.0005 BTC is not a fee that electrum would choose.)

**Comment by dabura667** [comment]
also, the minimum relay transaction fee is now 0.0001 per kb (which means 10 satoshis per byte) so only places like 21 etc. are going to relay those transactions that are less than 10 satoshis per byte.

**Comment by medoix** [comment]
I am the "user" and i used the Electrum Sweep function (clearly not many people know how this function works as everyone keeps telling me i choose the wrong fee) this function gives no options to modify the fee so no i did not "ballpark" it, electrum decided on the fee itself even after it knew the inputs etc.

**Comment by dabura667** [comment]
@medoix did you uncheck "Dynamic Fees" in your preferences?

I think the default if you uncheck Dynamic Fees is 0.0005 BTC per kB

**Comment by medoix** [comment]
@dabura667 that was indeed unchecked (although i never remember unchecking that). I have upgraded electrum numerous times on this PC. It is a bit late for me to test sweep transaction again as it has been confirmed now.

**Comment by medoix** [comment]
Could be worth adding in a function to notify the end user that the fee is low and therefore would take XX mins/hours/if at all based on the data at https://bitcoinfees.21.co/api

**Comment by gits7r** [comment]
Right, the sweep option should calculate the fee better or allow manual override. Also, not long ago we added notifications for incoming transactions which have fee under than recommended and may take longer time to confirm than usual. We should make it that it also warns for outgoing (sent) transactions before it allows signing and/or broadcasting and the same for the sweep function (if it will allow manual fee override). @ecdsa should this fee notification for outgoing transactions be opened as a new separate ticket? In 2.6.4 it calculates a fee of 0.000113 but I can manually override it to 0.000001 and don't get any warnings.

**Comment by gits7r** [comment]
It does not matter if he selected dynamic fees or not. the sweep function should
- calculate the fee properly by querying the server and getting the expected fee level in the network at that time, and choose a value which will get the transaction confirmed regardless of the fee settings;
- don't create a single monster sized transaction if the address being swept has many many unspent inputs, create multiple transactions instead;
- allow fee override (per each transactions if more are created due to too many unspent inputs for the address being swept) as well as warn if the fee is under recommended and confirmation might take longer than expected;
- if we sweep an address containing coinbase transaction (newly generated coins) wait for at least 'n' blocks on top of the block with that coinbase  -- this is already enforced as consensus rule in bitcoin, too new coins cannot be spent until some blocks are built on top.

**Comment by dabura667** [comment]
@gits7r May I link to this thread every time someone complains on another wallet about "not letting them choose fees"

Most of the time they say "You should give the users the ability to set static fees if they want to, freedom!"

Then cases like this happen and you say "You should automatically calculate the fees for the user regardless of their settings!"

I understand you are not the same person as the former, but still. People wonder why developers never bend to every whim asap... because 50 other people will pop up saying "hey put that back!"

-----------

In regards to your other points:

1. Single monster sized transaction: 5 kb is not "monster sized" by my opinion. And 5 kb has no bearing on the extent that it propagates across the network. The only time that raw bytesize would affect propagation adversely is somewhere over 50 kB. However, the default for isStandard() that decides whether to propagate a transaction or not, is set to 100kB, so any transaction that is valid and has a large enough fee will be relayed by bitcoind up to 100000 bytes. When you make a suggestion, please be more specific and explain why you chose the specific numbers. This will increase your chances of getting the feature included.
2. If we calculate the fee properly, the user will never need to be warned, and if the user doesn't trust our fees, then why should they trust the warning that it is under the recommendation. This doesn't make any sense.
3. Electrum already checks for coinbase wait period when returning utxos from the server. Please don't assume something is wrong, and report only things that you have documented as being wrong.

**Comment by medoix** [comment]
@dabura667 i agree, and i assume that feature was put in to appease more technical people (me included) and as i stated i never recall turning it off (no reason too) so not sure if it was disabled in the many years of upgrades i have been using Electrum for or i simply turned it off a long time ago and never had an issue pop up to remind me to turn it back on... not blaming anyone but as a user the simple fact was i had no idea why my transaction was not getting the correct fee.

In simple terms regarding the issue from my perspective it is "As a user i just want to know why my transaction may fail or is taking a long time to process" how that gets done is completely up to you devs.

Based on your number 2 above, if someone has chosen to set manual fees then at least the warning you refer to would advise them of this potential issue instead of the user just pulling hair out wondering why.

**Comment by gits7r** [comment]
@dabura667 I didn't say electrum doesn't already check for the coinbase period before returning utxos from the server, I just made a checklist with everything that needs to happen in the background, what is already happening should just be jumped over.

Secondly, even if you calculate the fee properly the option for manual override is still useful. Maybe I want a monster fee for that transaction. When sending normal transactions I can override the fee with a bigger one manually, but not in the sweep function. And of course if I override it with a too low fee, I should be warned that it might not get confirmed so yes both options are useful and have real use cases.

Third, it's not just the size of the transaction, it's the too many dust inputs in a single transaction. This is the reason why a node rejected the txid in this thread. So there's a limit (besides the kB limit you refer to) that limits the number of dust inputs in a single tx.

**Comment by dabura667** [comment]
> This is the reason why a node rejected the txid in this thread.

Which node, and how do you know that is the reason?

**Comment by gits7r** [comment]
This was also the error listed on bc.i -> Too many dust inputs, transaction was pruned from our node's mempool. I couldn't find it in my full node's mempool as well so I assume this is the reason why it was pruned from mine as well, given the fee was not 0 and everything else was valid. Irrelevant the reason, some improvements can be made to our sweep function as described above.

**Comment by purestmoney** [comment]
I register Multisig account on multisig for the purpose of business with a vendor, now after i have deposited money into the wallet, he sent me a fake document to show that he has done my job, but it was not done, and since he has one of the key, i cannot withdraw my money without the second key and i cannot pay him my money because he is a scam, i want admin to please tell me how i can withdraw my money, am ready to do anything am told to do please. This is a borrowed money, if lost then i will run on big loss.  Thanks   youcancomein2@yahoo.com

**Comment by dabura667** [comment]
@purestmoney This thread is not relevant to your issue. Please refrain from posting in other issues.

**Comment by purestmoney** [comment]
alright

**Comment by ecdsa** [comment]
I think the fee is miscalculated: it uses the fee_per_kb as fee, regardless of tx size

**Comment by ecdsa** [comment]
fixed, see 1e55f4fda078d7ff1882567101aab5f10c8b9772 and 5d44474aeb595ba4fd037188bc41b616825498a0

**Comment by ecdsa** [comment]
unfortunately blockchain.address.listunspent does not say if the input is coinbase, so we cannot filter out unmatured inputs with that function
